From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mike Hommey <mh+mozilla@glandium.org>
Date: Thu, 18 Jul 2024 21:07:44 +0000
Subject: [PATCH] Bug 1908637 - Allow configure options with at most 1 argument
 to contain commas. r?build

Comma is the separator for values in options. Like in --with-foo=bar,baz.
This is meant to be useful for options with nargs="*", "+", or a fixed
value greater than 1.

Currently, when an option with another nargs is passed a comma, we
reject the value as being multiple values, while one or less is
expected. This causes problems with options with nargs=1 that actually
CAN have legitimate values containing commas. Examples include:
  CFLAGS="-Wp,--preprocessor-flag -Wa,--assembler-flag"
  LDFLAGS=-Wl,--linker-flag"

Since options with multiple arguments are the exception, not the norm,
let's allow commas in cases where at most a single argument is expected.

An alternative would be to explicitly disable comma-splitting on the set
of options that can take values containing commas, but that's error-prone.
Differential Diff: PHID-DIFF-yh3xrdry2seh3abeobyd
---
 python/mozbuild/mozbuild/configure/options.py | 22 ++++++---
 .../mozbuild/test/configure/test_options.py   | 47 ++++++++++---------
 2 files changed, 40 insertions(+), 29 deletions(-)

diff --git a/python/mozbuild/mozbuild/configure/options.py b/python/mozbuild/mozbuild/configure/options.py
index 874f4cad740a..293f59bd9899 100644
--- a/python/mozbuild/mozbuild/configure/options.py
+++ b/python/mozbuild/mozbuild/configure/options.py
@@ -340,20 +340,27 @@ class Option(object):
         self.category = category or _infer_option_category(define_depth)
 
     @staticmethod
-    def split_option(option):
+    def split_option(option, values_separator=","):
         """Split a flag or variable into a prefix, a name and values
 
         Variables come in the form NAME=values (no prefix).
         Flags come in the form --name=values or --prefix-name=values
         where prefix is one of 'with', 'without', 'enable' or 'disable'.
-        The '=values' part is optional. Values are separated with commas.
+        The '=values' part is optional. Values are separated with
+        `values_separator`. If `values_separator` is None, there is at
+        most one value.
         """
         if not isinstance(option, six.string_types):
             raise InvalidOptionError("Option must be a string")
 
-        elements = option.split("=", 1)
-        name = elements[0]
-        values = tuple(elements[1].split(",")) if len(elements) == 2 else ()
+        name, eq, values = option.partition("=")
+        if eq:
+            if values_separator is None:
+                values = (values,)
+            else:
+                values = tuple(values.split(values_separator))
+        else:
+            values = ()
         if name.startswith("--"):
             name = name[2:]
             if not name.islower():
@@ -429,7 +436,10 @@ class Option(object):
                 % (option, origin, ", ".join(self.possible_origins))
             )
 
-        prefix, name, values = self.split_option(option)
+        kwargs = {}
+        if self.maxargs <= 1:
+            kwargs["values_separator"] = None
+        prefix, name, values = self.split_option(option, **kwargs)
         option = self._join_option(prefix, name)
 
         assert name in (self.name, self.env)
diff --git a/python/mozbuild/mozbuild/test/configure/test_options.py b/python/mozbuild/mozbuild/test/configure/test_options.py
index 91f296802328..b88adf1a2fb0 100644
--- a/python/mozbuild/mozbuild/test/configure/test_options.py
+++ b/python/mozbuild/mozbuild/test/configure/test_options.py
@@ -443,21 +443,22 @@ class TestOption(unittest.TestCase):
                     str(e.exception), "--%s takes %d values" % (name, nargs)
                 )
 
-        if nargs in (2, "*", "+") and not disabled:
+        if nargs in (1, "?", 2, "*", "+") and not disabled:
             value = option.get_value("--%s=foo,bar" % name, "option")
-            self.assertEqual(value, PositiveOptionValue(("foo", "bar")))
+            if nargs in (1, "?"):
+                self.assertEqual(value, PositiveOptionValue(("foo,bar",)))
+            else:
+                self.assertEqual(value, PositiveOptionValue(("foo", "bar")))
             self.assertEqual(value.origin, "option")
         else:
             with self.assertRaises(InvalidOptionError) as e:
                 option.get_value("--%s=foo,bar" % name, "option")
             if disabled:
                 self.assertEqual(str(e.exception), "Cannot pass a value to --%s" % name)
-            elif nargs == "?":
-                self.assertEqual(str(e.exception), "--%s takes 0 or 1 values" % name)
             else:
                 self.assertEqual(
                     str(e.exception),
-                    "--%s takes %d value%s" % (name, nargs, "s" if nargs != 1 else ""),
+                    "--%s takes %d values" % (name, nargs),
                 )
 
         option = Option("--%s" % name, env="MOZ_OPTION", nargs=nargs, default=default)
@@ -529,20 +530,20 @@ class TestOption(unittest.TestCase):
                 option.get_value("MOZ_OPTION=foo", "environment")
             self.assertEqual(str(e.exception), "MOZ_OPTION takes %d values" % nargs)
 
-        if nargs in (2, "*", "+"):
+        if nargs in (1, "?", 2, "*", "+"):
             value = option.get_value("MOZ_OPTION=foo,bar", "environment")
-            self.assertEqual(value, PositiveOptionValue(("foo", "bar")))
+            if nargs in (1, "?"):
+                self.assertEqual(value, PositiveOptionValue(("foo,bar",)))
+            else:
+                self.assertEqual(value, PositiveOptionValue(("foo", "bar")))
             self.assertEqual(value.origin, "environment")
         else:
             with self.assertRaises(InvalidOptionError) as e:
                 option.get_value("MOZ_OPTION=foo,bar", "environment")
-            if nargs == "?":
-                self.assertEqual(str(e.exception), "MOZ_OPTION takes 0 or 1 values")
-            else:
-                self.assertEqual(
-                    str(e.exception),
-                    "MOZ_OPTION takes %d value%s" % (nargs, "s" if nargs != 1 else ""),
-                )
+            self.assertEqual(
+                str(e.exception),
+                "MOZ_OPTION takes %d values" % nargs,
+            )
 
         if disabled:
             return option
@@ -588,20 +589,20 @@ class TestOption(unittest.TestCase):
                 env_option.get_value("MOZ_OPTION=foo", "environment")
             self.assertEqual(str(e.exception), "MOZ_OPTION takes %d values" % nargs)
 
-        if nargs in (2, "*", "+"):
+        if nargs in (1, "?", 2, "*", "+"):
             value = env_option.get_value("MOZ_OPTION=foo,bar", "environment")
-            self.assertEqual(value, PositiveOptionValue(("foo", "bar")))
+            if nargs in (1, "?"):
+                self.assertEqual(value, PositiveOptionValue(("foo,bar",)))
+            else:
+                self.assertEqual(value, PositiveOptionValue(("foo", "bar")))
             self.assertEqual(value.origin, "environment")
         else:
             with self.assertRaises(InvalidOptionError) as e:
                 env_option.get_value("MOZ_OPTION=foo,bar", "environment")
-            if nargs == "?":
-                self.assertEqual(str(e.exception), "MOZ_OPTION takes 0 or 1 values")
-            else:
-                self.assertEqual(
-                    str(e.exception),
-                    "MOZ_OPTION takes %d value%s" % (nargs, "s" if nargs != 1 else ""),
-                )
+            self.assertEqual(
+                str(e.exception),
+                "MOZ_OPTION takes %d values" % nargs,
+            )
 
         return option
 
